src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/BalanceCalculator.java

```diff
@@ -0,0 +1,167 @@
+package uk.gov.homeoffice.digital.sas.balancecalculator;
+
+import com.google.common.collect.Range;
+import java.math.BigDecimal;
+import java.time.Duration;
+import java.time.LocalDate;
+import java.time.ZonedDateTime;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import uk.gov.homeoffice.digital.sas.balancecalculator.client.RestClient;
+import uk.gov.homeoffice.digital.sas.balancecalculator.models.accrual.Accrual;
+import uk.gov.homeoffice.digital.sas.balancecalculator.models.accrual.Agreement;
+import uk.gov.homeoffice.digital.sas.balancecalculator.models.accrual.Contributions;
+import uk.gov.homeoffice.digital.sas.balancecalculator.models.accrual.enums.AccrualType;
+import uk.gov.homeoffice.digital.sas.balancecalculator.models.timecard.TimeEntry;
+
+@Component
+public class BalanceCalculator {
+
+  private final RestClient restClient;
+
+  @Autowired
+  public BalanceCalculator(RestClient restClient) {
+    this.restClient = restClient;
+  }
+
+  public List<Accrual> calculate(TimeEntry timeEntry) {
+
+    List<Accrual> accrualsToBatchUpdate = Arrays.stream(AccrualType.values())
+        .filter(AccrualType::isEnabled)
+        .map(accrualType -> calculateAccruals(timeEntry, accrualType))
+        .flatMap(Collection::stream)
+        .toList();
+
+    // TODO : send Batch Update request to Accruals API
+    // restClient.batchUpdate(accrualsToBatchUpdate)
+
+    return accrualsToBatchUpdate;
+  }
+
+  List<Accrual> calculateAccruals(TimeEntry timeEntry, AccrualType accrualType) {
+
+    Map<LocalDate, Range<ZonedDateTime>> dateRangeMap =
+        splitOverDays(timeEntry.getActualStartTime(), timeEntry.getActualEndTime());
+
+    return dateRangeMap.entrySet().stream()
+        .map(entry -> recalculateContributions(timeEntry.getId(), timeEntry.getTenantId(),
+            timeEntry.getOwnerId(), accrualType, entry.getKey(), entry.getValue()))
+        .map(accrual -> cascadeCumulativeTotal(timeEntry, accrual)).flatMap(Collection::stream)
```

Comment:
Although I know there was a decision to not deal with time entries spanning days I think this is on a path to need more rework. The cascadeCumulativeTotal function retrieves accruals, it feels like this responsibility should be split from the cascade. i.e. First retrieve all of the affected accruals then pass them into the `recalculateContributions` and `cascadeCumulativeTotal` methods
e.g. 
```
.map(entry -> getAccrualsFromReferenceDateUntilEndOfAgreement(dateRangeMap))
.map(accruals -> recalculateContributions(accruals, .... )
```